### ISSUE TITLE
Faster u:

### ISSUE THREAD

**Issue by moon-chilled** [jsoftware/jsource#157]
https://github.com/simdutf/simdutf (& 'clausecker' branch)

**Comment by clausecker** [comment]
@bilam I have assembly implementations of the AVX-512 code (clausecker branch) and can write assembly implementations for AVX2 if desired.  These are freestanding and easy to implement (plus they are faster!)

**Comment by bilam** [comment]
This is interesting. How much faster is it?

**Comment by clausecker** [comment]
@bilam See our upcoming paper for AVX-512: https://arxiv.org/abs/2212.05098

We reach 20 GB/s for ASCII, up to 10 GB/s for non-ASCII.

**Comment by bilam** [comment]
This j source guthub repo is a mirrored repo so that it can't merge PR. Please let me know what to do to incorporate your assembly code.

**Comment by clausecker** [comment]
I'm not at all familiar with your code base.  For the AVX-512 code, the relevant files are [`utf16le_to_utf8_avx512.S`](https://github.com/simdutf/simdutf/blob/clausecker/utf16le_to_utf8_avx512.S) and [`utf8_to_utf16le_avx512.S`](https://github.com/simdutf/simdutf/blob/clausecker/utf8_to_utf16le_avx512.S).  These provide one function each and need to be hooked up so they are only called when the CPU supports all required CPU features (which is AVX-512 VBMI2).

**Comment by bilam** [comment]
Which platforms are these assembly codes applicable?

**Comment by clausecker** [comment]
@bilam amd64 SysV ABI, which is most amd64 systems except Windows.  The code can be adapted to run on Windows if desired.  Windows has a different ABI requiring changes to the code.

**Comment by moon-chilled** [comment]
We can just declare `__attribute__((sysv_abi))` from the c side, rather than changing the assembly.  That's what I did for binary search too.  I forgot about the vbmi2 requirement, though; je avx512 baseline is skx, since that's the hardware I have, and skx doesn't support vbmi2.

**Comment by bilam** [comment]
I saw there are also implementation in c files using intrinsic. If their efficiency is roughly the same as assembly files, then I would prefer c with intrinsic over assembly. What's your opinion?

Btw is there also an avx2 implementation?

**Comment by HenryHRich** [comment]
I would very much prefer that we have no pure assembler, and use 
intrinsics only.

Henry Rich

**Comment by clausecker** [comment]
@bilam The C source is about 20% slower as the compiler has trouble with instruction scheduling and register allocation.  The C code is also not up to date with the latest code changes, but I could fix that.

A different algorithm is required for AVX2 as it lacks many of the features I used in the AVX-512 kernel.  There are publications on such algorithms and Lemire's library has C++ code implementing them (though it's template-ladden and not trivial to convert to pure C).  I have only written the AVX-512 variant.

**Comment by bilam** [comment]
@HenryHRich the assembly codes don't work for windows, although I think it just needs some changes in prolog/epilog. 

@clausecker is the assembly code generated from perl?  If so, can it target for nasm format or windows abi? Just curious.

**Comment by clausecker** [comment]
@clausecker No, it's hand written.  To make it work on Windows, the easiest way is to just declare the function as having SysV ABI, as @moon-chilled suggested.  I can add some preprocessor macros to allow it to assemble with either ABI if you like.

**Comment by bilam** [comment]
Personally I would like a macro solution that works for both *nix and windows.

**Comment by clausecker** [comment]
I can also write assembly implementations for AVX2 (and possibly SSE if desired), but as that's a considerable amount of work, I cannot do it for free.

Will see about macro use for Windows tomorrow.

**Comment by moon-chilled** [comment]
I would rather set the calling convention from the c side than use assembly macros.  No conditional compilation, no moving parts; same code works trivially everywhere.

**Comment by clausecker** [comment]
@bilam ABI adapters added.

**Comment by bilam** [comment]
Thanks. How do you estimate the length of output buffer?

**Comment by clausecker** [comment]
@bilam For UTF-16 → UTF-8, the worst case is characters in the range from U+0800 to U+FFFF which occupy one word in UTF-16 and three bytes of UTF-8.  So the word count is multiplied by 3 to get a byte count.  For UTF-8 → UTF-16, the worst case is ASCII where 1 byte is transcoded into 1 UTF-16 word, so the byte count is the word count.

**Comment by bilam** [comment]
I have added your assembly code into the repos. However there is no hardware capable of vbmi2 to test the code.

**Comment by clausecker** [comment]
@bilam Cool!  It's a bit unfortunate.  A processor of Icelake or newer generation is required.  AMD Zen 4 works, too.

**Comment by HenryHRich** [comment]
This code was put into JE some time ago
